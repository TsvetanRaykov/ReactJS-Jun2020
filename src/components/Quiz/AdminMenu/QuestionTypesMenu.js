import React from 'react'
import Button from '@material-ui/core/Button'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'

export default function SimpleMenu(props) {
	const { classes, disabled, handleQuestionTypeSelect } = props
	const [anchorEl, setAnchorEl] = React.useState(null)

	const handleClick = (event) => {
		setAnchorEl(event.currentTarget)
	}

	const handleClose = (type) => {
		if (['single', 'multiple', 'open'].includes(type)) {
			handleQuestionTypeSelect(type)
		}
		setAnchorEl(null)
	}

	return (
		<div>
			<Button
				className={classes.green}
				color='primary'
				variant='outlined'
				onClick={handleClick}
				disabled={disabled}
			>
				Add Question
			</Button>
			<Menu
				id='new-question-type-menu'
				anchorEl={anchorEl}
				keepMounted
				open={Boolean(anchorEl)}
				onClose={handleClose}
			>
				<MenuItem onClick={() => handleClose('single')}>Single answer</MenuItem>
				<MenuItem onClick={() => handleClose('multiple')}>
					Multiple answers
				</MenuItem>
				<MenuItem onClick={() => handleClose('open')}>Open answer</MenuItem>
			</Menu>
		</div>
	)
}
