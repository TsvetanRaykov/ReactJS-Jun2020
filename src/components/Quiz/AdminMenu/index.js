import React, { useContext, useState, useEffect } from 'react'

import { Paper, Button, Box, makeStyles } from '@material-ui/core'
import { Link, withRouter } from 'react-router-dom'
import Context from '../../../Context'
import quizService from '../../../services/quizService'
import { deepEqual } from '../../../utils'
import Green from '@material-ui/core/colors/green'
import ModalDialog from '../../shared/ModalDialog'
import QuestionTypesMenu from '../AdminMenu/QuestionTypesMenu'

const useStyles = makeStyles((theme) => ({
	green: {
		backgroundColor: Green[500],
		color: theme.palette.primary.contrastText,
		'&:hover': {
			color: theme.palette.primary.dark,
		},
		'&:disabled': {
			boxShadow: 'none',
			background: 'rgb(255, 255, 255)',
			color: 'rgba(0, 0, 0, 0.26)',
		},
	},
}))

const EditMenu = (props) => {
	const classes = useStyles()

	const { formHandler, isFormOpen, history } = props
	const { quiz, quizSnapshot } = useContext(Context)
	const [canSave, setCanSave] = useState(false)
	const [notSaved, setNotSaved] = useState(false)
	const [modalDialog, setModalDialog] = useState({
		title: '',
		message: '',
		open: false,
		handleYes: () => {},
		handleNo: () => {},
	})

	history.block(function ({ pathname }) {
		if (pathname.startsWith('/quiz/edit')) {
			return true
		}
		if (notSaved) {
			setModalDialog(() => {
				return {
					title: 'There are unsaved changes. Are you sure you want to leave?',
					open: true,
					handleYes: () => {
						setModalDialog({ open: false })
						setNotSaved(false)
						setTimeout(() => history.push(pathname), 0)
					},
					handleNo: () => {
						setModalDialog(() => ({
							open: false,
						}))
					},
				}
			})
			return false
		}
	})

	useEffect(() => {
		const errors = { questions: '' }

		errors.questions =
			!quiz?.questions || quiz.questions.length === 0
				? 'The Quiz must have at least one question.'
				: ''
		const hasNoErrors = !Object.entries(errors).find(([k, v]) => v.length > 0)
		const hasChanges = !deepEqual(quiz, quizSnapshot)
		setNotSaved(hasChanges)
		setCanSave(hasNoErrors && hasChanges && !isFormOpen)
	}, [isFormOpen, quiz, quizSnapshot])

	// const addQuestionHandler = () => {
	// 	formHandler(true, true)
	// }

	const saveQuizHandler = () => {
		setModalDialog(() => {
			return {
				title: 'Ready to save and go?',
				open: true,
				handleYes: () => {
					setModalDialog({ open: false })
					setNotSaved(false)
					setTimeout(() => {
						quizService.setQuiz(quiz)
						history.push('/dashboard')
					}, 0)
				},
				handleNo: () => {
					setModalDialog(() => ({
						open: false,
					}))
				},
			}
		})
	}

	const handleQuestionTypeSelect = (type) => {
		formHandler(true, type)
	}

	return (
		<>
			<Paper>
				<Box
					display='flex'
					justifyContent='space-between'
					flexDirection='row'
					p={1}
				>
					<Button
						color='primary'
						variant='outlined'
						component={Link}
						to='/quiz/edit'
					>
						Edit Details
					</Button>
					{/* <Button
						className={classes.green}
						color='primary'
						variant='outlined'
						onClick={addQuestionHandler}
					>
						Add Question
					</Button> */}
					<QuestionTypesMenu
						classes={classes}
						disabled={isFormOpen}
						handleQuestionTypeSelect={handleQuestionTypeSelect}
					/>
					<Button
						color='primary'
						variant={notSaved ? 'contained' : 'outlined'}
						onClick={saveQuizHandler}
						disabled={!canSave}
					>
						Save Quiz
					</Button>
				</Box>
			</Paper>
			<ModalDialog {...modalDialog} />
		</>
	)
}

export default withRouter(EditMenu)
