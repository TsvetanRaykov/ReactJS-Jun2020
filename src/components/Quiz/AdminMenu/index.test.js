import React from 'react'
import renderer from 'react-test-renderer'
import TestWrapper from '../../../utils/test-wrapper'
import AdminMenu from '.'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<AdminMenu />', () => {
	it('Should match the snapshot', () => {
		const wrapper = renderer.create(
			<TestWrapper>
				<AdminMenu />
			</TestWrapper>
		)
		expect(wrapper).toMatchSnapshot()
	})
})
