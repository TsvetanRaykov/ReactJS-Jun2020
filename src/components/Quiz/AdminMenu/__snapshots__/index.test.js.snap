// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminMenu /> Should match the snapshot 1`] = `
Array [
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiBox-root MuiBox-root-2"
    >
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        href="/quiz/edit"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="MuiButton-label"
        >
          Edit Details
        </span>
      </a>
      <div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-green-1 MuiButton-outlinedPrimary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Add Question
          </span>
        </button>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Save Quiz
        </span>
      </button>
    </div>
  </div>,
  <div />,
]
`;
